pkg/storage/fs/ocis/path.go

```diff
@@ -0,0 +1,189 @@
+package ocis
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	userpb "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	"github.com/cs3org/reva/pkg/errtypes"
+	"github.com/cs3org/reva/pkg/storage/utils/templates"
+	"github.com/pkg/errors"
+)
+
+// Path implements transformations from filepath to node and back
+type Path struct {
+	// ocis fs works on top of a dir of uuid nodes
+	Root string `mapstructure:"root"`
+
+	// UserLayout wraps the internal path with user information.
+	// Example: if conf.Namespace is /ocis/user and received path is /docs
+	// and the UserLayout is {{.Username}} the internal path will be:
+	// /ocis/user/<username>/docs
+	UserLayout string `mapstructure:"user_layout"`
+
+	// EnableHome enables the creation of home directories.
+	EnableHome bool `mapstructure:"enable_home"`
+}
+
+// Resolve takes in a request path or request id and converts it to a NodeInfo
+func (pw *Path) Resolve(ctx context.Context, ref *provider.Reference) (*NodeInfo, error) {
+	if ref.GetPath() != "" {
+		return pw.Wrap(ctx, ref.GetPath())
+	}
+
+	if ref.GetId() != nil {
+		return pw.WrapID(ctx, ref.GetId())
+	}
+
+	// reference is invalid
+	return nil, fmt.Errorf("invalid reference %+v", ref)
+}
+
+// Wrap converts a filename into a NodeInfo
+func (pw *Path) Wrap(ctx context.Context, fn string) (node *NodeInfo, err error) {
+	var link, root string
+	if fn == "" {
+		fn = "/"
+	}
+	if pw.EnableHome && pw.UserLayout != "" {
+		// start at the users root node
+		var u *userpb.User
+
+		u, err = getUser(ctx)
+		if err != nil {
+			err = errors.Wrap(err, "ocisfs: Wrap: no user in ctx and home is enabled")
+			return
+		}
+
+		layout := templates.WithUser(u, pw.UserLayout)
+		root = filepath.Join(pw.Root, "users", layout)
+
+	} else {
+		// start at the storage root node
+		root = filepath.Join(pw.Root, "nodes/root")
+	}
+
+	node, err = pw.ReadRootLink(root)
+	// The symlink contains the nodeID
+	if err != nil {
+		return
+	}
+
+	if fn != "/" {
+		// we need to walk the path
+		segments := strings.Split(strings.TrimLeft(fn, "/"), "/")
+		for i := range segments {
+			node.ParentID = node.ID
```

Comment:
maybe extract these bits into a function with a good name to make it clearer what is happening and why